src/Nethermind/Nethermind.Consensus/Processing/ParallelProcessing/ParallelRunner.cs

```diff
@@ -0,0 +1,84 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Threading;
+
+namespace Nethermind.Consensus.Processing.ParallelProcessing;
+
+public class ParallelRunner<TLocation, TLogger>(
+    ParallelScheduler<TLogger> scheduler,
+    MultiVersionMemory<TLocation, TLogger> memory,
+    ParallelTrace<IsTracing> parallelTrace,
+    IVm<TLocation> vm,
+    int? concurrencyLevel = null) where TLogger : struct, IIsTracing where TLocation : notnull
+{
+    private int _threadIndex = -1;
+
+    public async Task Run()
+    {
+        int concurrency = concurrencyLevel ?? Environment.ProcessorCount;
+        using ArrayPoolList<Task> tasks = new ArrayPoolList<Task>(concurrency);
+        for (int i = 0; i < concurrency; i++)
+        {
+            tasks.Add(Task.Run(Loop));
+        }
+
+        await Task.WhenAll(tasks.AsSpan());
+    }
+
+    private async Task Loop()
+    {
+        int threadIndex = Interlocked.Increment(ref _threadIndex);
+        using var handle = Thread.CurrentThread.BoostPriorityHighest();
+        TxTask task = scheduler.NextTask();
+        do
+        {
+            if (task.IsEmpty)
+            {
+                // Alternatives:
+                // - Use ManualResetEventSlim and block threads, but don't have this complicated TCS stuff
+                // - Spin this loop until scheduler.Done (what paper proposes)
+                await scheduler.WorkAvailable.WaitAsync();
```

Comment:
This is questionable, other mentioned approaches can be found:
- ManualResetEventSlim: [d21026c](https://github.com/NethermindEth/nethermind/pull/8550/commits/d21026cbf5b944d864c12fe47075e95007d90639) and later fix in [0710541](https://github.com/NethermindEth/nethermind/pull/8550/commits/071054173838db6a845e166da9c2877ca70db8ba)
- Spinning: [b1e015f](https://github.com/NethermindEth/nethermind/pull/8550/commits/b1e015ffecfe6fd172f2bc0fde57062e7adbaa9a)

I'm happy to hear more about your thoughts